task-worklow-refactor/README.md

```diff
@@ -0,0 +1,94 @@
+# How to split one Celery task into multiple tasks
+
+Let's talk about splitting the `task.process_message` into one processing task
+and multiple tasks handling the specific events. The main benefit from this 
+will be the possibility to use more queues (for faster/slower tasks)
+ and to differentiate tasks easily (Sentry).
+
+
+## Workflow
+ - get webhook payload or message from message bus 
+ - add aditional info which can make determining the event type easier
+ (GH webhooks - X-GitHub-Event)
+ - send it to Celery as `task.process_message` - this task will take care about:
+    - parsing of the event into object of class Event (this class provides `project`, `package_config` properties,
+     which enables us to make the checks)
+    - private repository check
+    - getting the handlers which handle the event
+    - whitelist check
+    - possibly creating records in DB (explained below)
+    - (running `pre-check` here to filter out some events?)
+    - sending specific task to Celery
+  - handle specific tasks, which directly run the handlers
+    
+#### How to pass the information needed by handlers
+ Handlers are currently using the information from event (with `project` and `package_config` properties, specific for each event),
+  service config and job config object:
+ ```python
+class JobHandler(Handler):
+    def __init__(
+        self, config: ServiceConfig, job_config: Optional[JobConfig], event: Event
+    )
+```
+ Possible solutions, which can be somehow combined:
+ 1. serialize the info about objects and pass it into `send_task`
+     - this would need serializing and then again deserializing
+ 2. save the info about objects in DB and pass IDs of models into `send_task`
+     - what models does make sense to have? possibilities:
+         - project
+         - package config 
+         - service config
+         - job config 
+         - event
```

Comment:
Most of them can be calculated from the attributes of the event. => We can save only those attributes, that are needed for initialization of the event class. (And the information about event type.)

Saving and recreating the event instances is probably not the cleanest way how to do it.

Reply:
- `config`, `package_config` and `job_config` can be easily serialize
- `project` can be saved to the database and instantiate once again from the `config`. (We can forward database id.)
- `event` will cause more problems, it contains some metadata about the event itself (used in handlers) and we are using its type during the workflow to differentiate the responses.